### PR TITLE
Update DotSet expression to always take a ChainExpr on the LHS

### LINKED ISSUE #2000: Bad format result with SynExpr.DotSet
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAXAngB0WAQgIYBOAFAJRgC8AOsTQHYQQC2iLARosWOgBYBLAM4A6AGaxYYAO4D%2BYAOaJ0YCtTAMArgBttYQg3hghy1QDdC2zYhSwOAK0pVV5Oo2Zg3DbacLZsLmBoRDzOIRReXn7YFOKSYAA8ALRgwAA%2BXh4AGuoALACsmcwAmtRFHhAA9JVgcCxsDOjlFQDsAJzN1WAAXr3dzQBa6gBMAMwADF5pAL6QADTgUCbo6AIMCkKoYADaixCgFVAAJFi4W5Br6PN7zJBHFlbIaLuHEOMLr1AAkoaIjQDKAm6yA%2Br1yNwgAF0btNQfsIcdTk8oJdroc7g9rFsXq8AIxwiqQACyhAAHgAZNaIcl-BT8NF44bjBHQjywm4HQknHDIyAcSTaBkeDGWLHPBEQYYE4X-dgCODaWAMACC6AAooYAPJiSkMEESsBiSwmFkwgmc4Xcs5oPkCoW3e6i5E4w6jaW3f7YQjQRD4RASYiIAAKJEIbHQ3HtHnQxGsprZ5oRdyR535sEF7pFj2xBty7ogkE93t9-tggfJsGk3GghBMPzMsBrqyVUeYMbjh1ZzHZHgtDpTNrTGaTjuz4s%2B%2BXzUCLPr9AcQAFV-NXa4h643CM2GK2IEbtCbO2aOSOB1Ah62s2KdgaAGxTwte2elwMAYW0taEL6VQnb0HQZZ3Q1jUQeNu0TdErV5c9M1HK8XVae8ZxLeciXYLhiEAvcDwqLsIB7Zg%2BwLSDUztGDMWdA0AA5EMfZCy0QL9FW3KcsJAw8E2PCDT1tdML1gijPjaGji2VMQI2IL96kIQD2zYnCj17E8eRI3iyKdHNPlxd4DQfYs53o2UWHldMWxY4DQLw8CuW46CR3IjS8XxHSkNE8TDOMpiZNjOSPFwsB8PhLjlMHUi7PU8dGXvH4kEaTUGAAFWITAAHU5D4TDzPYsDOOs4Kz1CiD7Iiw5cTdZzaOVYhYE0QwABFEG0AQjPEryO3kjjFKC60UUaPiiuvTS8x0klSS%2BMR4r4P41X3RB-j4Mt0FS%2BB6SnXJmSyyycstbjUTUscBrxSdhrJH4xAEUlNVwYhNzLNVSWwQMhCEARTINfJ1va7LOty7qLl6va4INXE72O0kACVEDgYh4CWlbcw%2B3yFIIpTft2sL9vgjxcRae8RohqH4AAOU0ThuG1L4IxYTYp1xCz-Ks7a8sgFgdFWRq9QAMTLFhN3E%2BLlIBgS8WonT8bLeAiVZgR2cQLniB5lZI3vaA%2BFDP9uAAfVkZb0rpgLUhR3k0cK8KDpKoTQcq67ME1YhKR-WH0tWhHmD8-XCMRJnja5frMeYJlcbJK3CBtu3hHQYnSeIcnKepoG9YZ-smZZ7Q2apOWFb5gX0cBz5hicz5IGD0P7fQSXU%2Bl9Pud5pWdJVtXxK1gQdcgBOtqT1H-pzoXDmGKVQYANSdfA1ngNYFEdwDKJdqEkcCn6ja7k2MYNMZA9JDmar-F6GBHwxx8n522%2B%2BxnO6uQWHN7obC5G2rYHQABxZQ7oexAnp3w%2B3pnsA3cToibIKj7U2ftJRHRvlLGW%2BBFTQAANb4GurA5QQg4r-DDAxdMJNmIGlYsfZGXUoKAMtL7VeINC6E0QNISBygqx-H5rgeqZ0GByB3sqQwqEo5x0%2BDgja9N27-yZrZZeudXjDBxjpAA0ogRA2AxoTT%2BD8VBbBdT6i4ZlT6m0T4d0Xufbul8KjDBFjfMk00jJCD4J-T4a1cHz1PgQ1Sujir6ItoXQE6wfCVQUCTP4i1EAcCJAgFRrxuHqN4Zo-hv1BFAJXp8CY95lSNQUAwTeDBt5KkBIkzcmhAzxVgNFbxm4d4ZRmtYg2%2BCVLDiET3CoowC6vCLqnbgDAClmGpEqBQ7D0IMLWMw78RTsKIw6nghe5S%2BrAINKMfu4CK7KPJGGDg8BCBvlgM9dY5DKFUj6T5V2c9SnDJCvYypeiPCjDKoXWqwhCAcB8CY4QfB-iYEaGSTZJSPbJiZn8eA2plH-CwD4UZ0TXijGvnUjUnydQbPvNoMQrceHu0NiMi%2BjjjlgLqZI6ReTGg-HgQYFWzzYV-09hEwhDpiExNIXUqBBgYHKKEJVaqhhyE-kQBLCBVIX6PWer0qcskXnwv2RUqJwjXRiMLiEfSgZzlCGgMQJqaxNzMoXEwpU9VoDvmuluPFIS4VlP5f8oV1TDF1J%2BTKv8-ikCaoGd2RYkJQQXCEBzZ6WwcHTCAA)

#### Code

```fsharp
type Bar() =
    member this.foo with get () = null and set (value:obj) = ()
    
let app () : Bar = Bar()

app().foo <- {|
    X = 45
    Y =
        // comment
        79
    // zzzz
    Z = 230
|}
```
            

#### Result

```fsharp
type Bar() =
    member this.foo
        with get () = null
        and set (value: obj) = ()

let app () : Bar = Bar()

app ().foo <- {| X = 45
                 Y =
                  // comment
                  79
                 Z = 230 |}

```
            
#### Problem description

There are a lot of things wrong with this result:
- The comment above field `Z` is missing (trivia problem)
- The  `rhsExpr` should be indented and put on the next line
- There should not be a space between `app` and `()`, the code doesn't compile anymore.

#### Extra information

- [x] The formatted result breaks by code.
- [ ] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this.

#### Options

Fantomas 4.6 branch at 1/1/1990

Default Fantomas configuration

<sub>Did you know that you can ignore files when formatting from fantomas-tool or the FAKE targets by using a [.fantomasignore file](https://github.com/fsprojects/fantomas/blob/master/docs/Documentation.md#ignore-files-fantomasignore)?</sub>

### PR DESCRIPTION
**Pull request by josh-degraw**
This was mainly inspired by #2000. 

This fixes the problem where the LHS of a DotSet expression would sometimes be changed if a unit was involved which would result in invalid code.

It seemed like the root of the remaining problem on that issue was that `unit` nodes in the left side of a `DotSet` expression should never have a space inserted under any circumstance, and to me that meant the left-hand side of a `DotSet` expression should always be treated like a Chain.

Open to any and all feedback if there is a better way to accomplish this.

Closes #2000

### CODE REVIEW

**Review by nojaf — CHANGES_REQUESTED** [review]
Hi again 😊, excited to see you tackle problems in the space of the AST transformation!

I like the idea of considering the `identifierExpr` of a `SynExpr.DotSet` as a chain.

I do think we can achieve this in a much simpler way though.
I believe `ExprDotSetNode`can be removed altogether and we can re-use  `ExprSetNode` to map a `SynExpr.DotSet` to.

If we convert the `targetExpr` and the `longDotId` to a `SynExpr.DotGet`, that could be transformed into a chain by the natural flow.

Something like:
```fsharp
| SynExpr.DotSet(e1, synLongIdent, e2, _) ->
    let mDot =
        mkRange
            e1.Range.FileName
            (Position.mkPos e1.Range.EndLine (e1.Range.EndColumn + 1))
            (Position.mkPos synLongIdent.Range.StartLine (synLongIdent.Range.StartColumn - 1))

    let dotGet =
        SynExpr.DotGet(e1, mDot, synLongIdent, unionRanges e1.Range synLongIdent.Range)

    ExprSetNode(mkExpr creationAide dotGet, mkExpr creationAide e2, exprRange)
    |> Expr.Set
```

Everything will play out accordingly, the `identifierExpr` in `ExprSetNode` will be a `ChainExpr` and we only need to remove code in CodePrinter.

It also preserves `|ChainExpr|_|` and doesn't require it to blend with a slightly related concept.

The only downside of this is that we don't have the exact range of the `.`. In my code above, I made an approximate range. We could add that to the F# parser if we ever have a use-case to know its exact location.

File: src/Fantomas.Core/CodePrinter.fs [hunk 1]
```diff
@@ -1724,6 +1558,156 @@ let genExpr (e: Expr) =
     | Expr.IndexFromEnd node -> !- "^" +> genExpr node.Expr |> genNode node
     | Expr.Typar node -> genSingleTextNode node
 
+let genChain node =
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:1561]: I don't see much value in extracting the code from a match clause into a function when that function will only be called from one location (being the match clause body).

Ever since Dallas, I've found it easier to leave that code in the match clause body.

And again, make your PR diff as small as possible ;)

File: src/Fantomas.Core/ASTTransformer.fs [hunk 2]
```diff
@@ -653,7 +653,7 @@ let mkLinksFromFunctionName (mkLinkFromExpr: SynExpr -> LinkExpr) (functionName:
               yield (mkLongIdentExprFromSynIdent lastSynIdent |> mkLinkFromExpr) ]
     | e -> [ mkLinkFromExpr e ]
 
-let (|ChainExpr|_|) (e: SynExpr) : LinkExpr list option =
+let (|ChainExpr|_|) (e: SynExpr) : (LinkExpr list * SynExpr) option =
```
Comment (nojaf) [inline src/Fantomas.Core/ASTTransformer.fs:656]: Please revert this change.

File: src/Fantomas.Core/ASTTransformer.fs [hunk 3]
```diff
@@ -807,7 +807,7 @@ let (|ChainExpr|_|) (e: SynExpr) : LinkExpr list option =
     | SynExpr.DotGet _
     | SynExpr.TypeApp(expr = SynExpr.DotGet _)
     | SynExpr.DotIndexedGet(objectExpr = SynExpr.App(funcExpr = SynExpr.DotGet _) | SynExpr.DotGet _) ->
-        Some(visit e id)
+        Some(visit e id, e)
```
Comment (nojaf) [inline src/Fantomas.Core/ASTTransformer.fs:810]: Please revert.

File: src/Fantomas.Core/ASTTransformer.fs [hunk 4]
```diff
@@ -1117,7 +1117,7 @@ let mkExpr (creationAide: CreationAide) (e: SynExpr) : Expr =
         ExprIndexWithoutDotNode(mkExpr creationAide identifierExpr, mkExpr creationAide indexExpr, exprRange)
         |> Expr.IndexWithoutDot
 
-    | ChainExpr links ->
+    | ChainExpr(links, _) ->
```
Comment (nojaf) [inline src/Fantomas.Core/ASTTransformer.fs:1120]: Please revert.

**Review by nojaf — CHANGES_REQUESTED** [review]
Just a little cleanup left.
Could you also add an exact test for https://github.com/fsprojects/fantomas/issues/2000?
I would assume that will be solved as well.
Oh and a changelog entry please 😊.

File: src/Fantomas.Core/CodePrinter.fs (hunk 1 above)
Comment (josh-degraw) [reply src/Fantomas.Core/CodePrinter.fs:1561]: Good catch, I had that pulled out when I needed it more than once, forgot to put it back 🙃

Comment (nojaf) [reply src/Fantomas.Core/CodePrinter.fs:1561]: No worries, I typically revert the entire file in this kind of situation.
`git checkout <hash-on-main> --files src/Fantomas.Core/CodePrinter.fs`

**Review by nojaf — APPROVED** [review]
Great work!
